File: 3885-count-special-triplets/count-special-triplets.java
class Solution {
    private static final int MOD = (int)1e9 + 7;
    public int specialTriplets(int[] nums) {
        int n = nums.length;
        Map<Integer, Integer> l = new HashMap<>();
        Map<Integer, Integer> r = new HashMap<>();
        long max = -1;

        for (int num : nums) {
            r.put(num, r.getOrDefault(num, 0) + 1);
            max = Math.max(max, num);
        }
        long cnt = 0;
        for (int j = 0; j < n; j++) {
            int m = nums[j];
            r.put(m, r.get(m) - 1);

            int lCnt = l.getOrDefault(m * 2, 0);
            int rCnt =  r.getOrDefault(m * 2, 0);

            cnt = (cnt + (long) lCnt * rCnt) % MOD;

            l.put(m, l.getOrDefault(m, 0) + 1);
        }

        return (int) cnt;
    }
}

File: 3885-count-special-triplets/count-special-triplets.cpp
class Solution {
public:
    int specialTriplets(vector<int>& nums) {
        int n = nums.size();
        int MOD = 1000000007;
        unordered_map<int, int> leftCount;
        unordered_map<int, int> rightCount;


        for (int num : nums) {
            rightCount[num] = rightCount[num] + 1;
        }

        long long count = 0;

        for (int j = 0; j < n; j++) {
            rightCount[nums[j]] -= 1;

            long long left = leftCount[nums[j] * 2];
            long long right = rightCount[nums[j] * 2];

            count = (count + (left * right) % MOD)% MOD;

            leftCount[nums[j]] += 1;
        }

        return (int)count;
    }
};